### ISSUE TITLE
sharing mechanism is not clear at the moment

### ISSUE THREAD

**Issue by yegor256** [Nerodesk/nerodesk#191]
Our current implementation of file sharing mechanism is not clear for end users, and can't be clear. We're expecting a user to enter URN of a friend, in order to start sharing a file with that friend. But the user can't know this URN anyhow. Well, it can, but he/she has to be very professional computer user.

Here is my suggestion how to fix that. When a user wants to share a file, he/she enters an email of his/her friend, next to the file to share. We send a welcome email to the friend. In that email we give a unique URL. When the URL is clicked, we start sharing the file. Thus, they should not know each other's URNs and it's fast.

**Comment by k0r** [comment]
@yegor256 awesome, thanks. let's implement this.

**Comment by k0r** [comment]
@yegor256 file sharing and commenting could look very similar to netbout, and/or we could share the links directly to the origins of the implemented federated logins.

**Comment by k0r** [comment]
@yegor256 public file sharing: twitter, fb, g+, and private file sharing by email that utilizes a framework similar to netbout for access, comments, etc.

**Comment by k0r** [comment]
@yegor256 I would like to be involved with this process.

**Comment by karato** [comment]
@yegor256 I will ask someone to take care of this task soon

**Comment by karato** [comment]
@yegor256 thanks for this report, I added 15 mins to your account, in transaction `000-1b04fe6a`

**Comment by yegor256** [comment]
@karato this is urgent

**Comment by karato** [comment]
> @karato this is urgent

@yegor256 thanks, I added the "urgent" tag

**Comment by karato** [comment]
@longtimeago this is your task

**Comment by longtimeago** [comment]
@yegor256 What does the label `urgent` mean?  I can't find anything about it in the policy.
Does it reduce the time to implement?

**Comment by yegor256** [comment]
@longtimeago see http://www.teamed.io/policy.html#26

**Comment by longtimeago** [comment]
@yegor256 I'm not sure I clearly understand the use case. You said

> In that email we give a unique URL. When the URL is clicked, we start sharing the file.

Should that friend be registered in our system? If he is not, how can file will be shared with him?
When you said "start sharing" you mean get the link to download the file?

Could you please elaborate a bit the idea

**Comment by yegor256** [comment]
@longtimeago yes, the user has to be logged in, when he clicks the link. The link will looks like this:

```
http://www.nerodesk.com/share?file=test.txt&owner=urn:github:123
```

Using these two parameters and having the user logged in we can share the file, right?

We can also add some secret code to the URL, but this is for the future

**Comment by yegor256** [comment]
@longtimeago yes, we need a secret code, otherwise anyone would be able to click this link. should be something like this:

```
http://www.nerodesk.com/share?file=test.txt&owner=urn:github:123&code=aef54be5cf45
```

This `aef54be5cf45` is calculated on the server, using some encoding mechanism with a secret key. We can use what Takes already provides, for example `CcXOR`. To get the code we should use the owner URN, the file name and the email of a friend.

**Comment by yegor256** [comment]
@longtimeago any progress here?

**Comment by longtimeago** [comment]
@yegor256 not yet, sorry :(. Feel free to take it if it should be done in 1 day

**Comment by longtimeago** [comment]
@yegor256 Please, see the PR https://github.com/Nerodesk/nerodesk/pull/250 if it's not too late

**Comment by karato** [comment]
@longtimeago 8 days already, in 24 hours I will re-assign; -30 added to your rating, at the moment it is: [+4746](http://www.netbout.com/b/35024?open=rating)

**Comment by karato** [comment]
@longtimeago task is too long, stop working with it... -60 to your rating, your total score is [+4656](http://www.netbout.com/b/35024?open=rating)

**Comment by karato** [comment]
@caarlos0 do this task

**Comment by caarlos0** [comment]
@karato I'm a little out of time, can you please assign to someone else?

**Comment by k0r** [comment]
@yegor256 does this ticket require a developer? #264 puts the sharing mechanism into focus.

**Comment by k0r** [comment]
@yegor256 sharing with URNs is no longer required.

**Comment by yegor256** [comment]
@k0r this ticket is about sharing via email, and is needed. #264 is just about a front-end of a file, not a sharing mechanism

**Comment by k0r** [comment]
@yegor256 I disagree. #264 is basically the sharing mechanism. and sharing by email is an editable parameter not dependent on URNs.

**Comment by yegor256** [comment]
#264 will just create a web page where some information about a file will be rendered, like file size, age, type, etc. that's it. there is no functionality there, just a rendering of the data we already have.

sharing is working only through URN's. email sharing is just a convenient extension on top of existing sharing mechanism, which is based on URNs. the only information we have about a user is his URN. when we share a file, we create a link between two URNs. that's how sharing work.

**Comment by k0r** [comment]
@yegor256 it's up to the user to distribute bit.ly links that point to file information pages rendered by takes/tacit. URNs would be useful for controlling the access to an uploaded file.

**Comment by k0r** [comment]
@yegor256 what if an access token is emailed rather than sharing with URNs. then, the file information page for private files would be accessed with the emailed access token whereas #264 is for public files with bit.ly.

**Comment by yegor256** [comment]
@k0r sorry, I didn't understand the question

**Comment by k0r** [comment]
@yegor256 is there an alternative way to share private files without URNs?

**Comment by yegor256** [comment]
@k0r I still don't understand the question. or it's not a question? don't see a question mark :) I would appreciate if you elaborate a bit more

**Comment by k0r** [comment]
@yegor256 updated comment.

**Comment by yegor256** [comment]
@k0r in order to "share" my file with you we need to connect my account with yours. accounts are identified by URNs. so, there is no other way to connect our accounts, except through URNs

**Comment by karato** [comment]
> @karato I'm a little out of time, can you please assign to someone else?

@caarlos0 30 points was deducted from your rating

**Comment by karato** [comment]
> @karato I'm a little out of time, can you please assign to someone else?

@caarlos0 right, I'll try to find someone else for this task

**Comment by karato** [comment]
@darkled do this task

**Comment by k0r** [comment]
@yegor256 we should postpone this task for now. sharing mechanism for private files isn't concise enough for task development. also, because this task seems quite laborious, I think alternative methods for sharing private files should be explored.

**Comment by k0r** [comment]
@yegor256 sharing mechanism for private files isn't concise enough for task development.

**Comment by k0r** [comment]
@yegor256 in the description, `unique URL` method should be explained.

**Comment by k0r** [comment]
@yegor256 also, because this task seems quite laborious, I think alternative methods for sharing private files should be explored.

**Comment by yegor256** [comment]
@karato please postpone this one

**Comment by k0r** [comment]
@yegor256 how is the unique URL generated? will this require using a database to keep track of which URNs have access to the file?

**Comment by yegor256** [comment]
@k0r URL generation is explained in https://github.com/Nerodesk/nerodesk/issues/191#issuecomment-95726865

**Comment by k0r** [comment]
@yegor256 will this require using a database to keep track of which URNs have access to the file?

**Comment by yegor256** [comment]
@k0r no, no persistence (database) will be required

**Comment by k0r** [comment]
@yegor256 the description indicates that this is a solution to not having to know the other users URN. based on the ticket description, this method of sharing files privately is generating a link for every email address the user wants to share the file with. anyone with this link could access the file though, right?

**Comment by yegor256** [comment]
@k0r yes that's correct

**Comment by k0r** [comment]
@yegor256 thanks for the clarification.

**Comment by karato** [comment]
> @karato please postpone this one

@yegor256 sure, thanks, I added "postponed" tag to it

**Comment by karato** [comment]
> @karato please postpone this one

@yegor256 thanks, I removed the "urgent" tags

**Comment by karato** [comment]
> @karato please postpone this one

@yegor256 right, I will find someone else, no problem

**Comment by k0r** [comment]
@karato you might want to take a look at the labels again.

**Comment by karato** [comment]
@darkled this is your task

**Comment by k0r** [comment]
@yegor256 isn't this supposed to be postponed? why does @karato keep changing the status of this ticket?

**Comment by k0r** [comment]
@yegor256 this task is supposed to be postponed. I don't know why @karato keeps changing the labels.

**Comment by k0r** [comment]
@yegor256 thanks.

**Comment by k0r** [comment]
@yegor256 why are the labels getting changed like this?

**Comment by k0r** [comment]
@karato please, stop changing the labels.

**Comment by yegor256** [comment]
@k0r I fixed the tag

**Comment by karato** [comment]
@k0r @yegor256 sorry about the mess with labels, my bad
